server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java

```diff
@@ -0,0 +1,429 @@
+            publication.onFaultyNode(n2);
+        }
+
+        assertTrue(publication.completed);
+        assertTrue(publication.committed);
+
+        publication.onFaultyNode(randomFrom(n1, n3)); // has no influence
+
+        List<Tuple<DiscoveryNode, Throwable>> errors = ackListener.awaitErrors(0L, TimeUnit.SECONDS);
+        assertThat(errors.size(), equalTo(1));
+        assertThat(errors.get(0).v1(), equalTo(n2));
+        assertThat(errors.get(0).v2().getMessage(), containsString("faulty node"));
+    }
+
+    public void testClusterStatePublishingFailsOrTimesOutBeforeCommit() throws InterruptedException {
+        VotingConfiguration config = new VotingConfiguration(Sets.newHashSet(n1.getId(), n2.getId()));
+        initializeCluster(config);
+
+        AssertingAckListener ackListener = new AssertingAckListener(nodes.size());
+        DiscoveryNodes discoveryNodes = DiscoveryNodes.builder().add(n1).add(n2).add(n3).localNodeId(n1.getId()).build();
+        MockPublication publication = node1.publish(CoordinationStateTests.clusterState(1L, 2L,
+            discoveryNodes, config, config, 42L), ackListener, Collections.emptySet());
+
+        boolean timeOut = randomBoolean();
+        publication.pendingPublications.entrySet().stream().collect(shuffle()).forEach(e -> {
+            if (e.getKey().equals(n2)) {
+                if (timeOut) {
+                    publication.onTimeout();
+                } else {
+                    e.getValue().onFailure(new TransportException(new Exception("dummy failure")));
+                }
+                assertTrue(publication.completed);
+                assertFalse(publication.committed);
+            } else if (randomBoolean()) {
+                PublishResponse publishResponse = nodeResolver.apply(e.getKey()).coordinationState.handlePublishRequest(
+                    publication.publishRequest);
+                e.getValue().onResponse(new PublishWithJoinResponse(publishResponse, Optional.empty()));
+            }
+        });
+
+        assertThat(publication.pendingCommits.keySet(), equalTo(Collections.emptySet()));
+        assertNull(publication.applyCommit);
+        assertTrue(publication.completed);
+        assertFalse(publication.committed);
+
+        List<Tuple<DiscoveryNode, Throwable>> errors = ackListener.awaitErrors(0L, TimeUnit.SECONDS);
+        assertThat(errors.size(), equalTo(3));
+        assertThat(errors.stream().map(Tuple::v1).collect(Collectors.toList()), containsInAnyOrder(n1, n2, n3));
+        errors.stream().forEach(tuple ->
+            assertThat(tuple.v2().getMessage(), containsString(timeOut ? "timed out" :
+                tuple.v1().equals(n2) ? "dummy failure" : "non-failed nodes do not form a quorum")));
+    }
+
+    public void testClusterStatePublishingTimesOutAfterCommit() throws InterruptedException {
+        VotingConfiguration config = new VotingConfiguration(Sets.newHashSet(n1.getId(), n2.getId()));
```

Comment:
Could we (randomly?) add `n3` to the configuration here?

Reply:
sure, fixed in 799157d72d4